elytron/src/main/java/org/wildfly/extension/elytron/HttpAuthenticationFactoryAddHandler.java

```diff
@@ -0,0 +1,97 @@
+ */
+
+abstract class HttpAuthenticationFactoryAddHandler<T> extends BaseAddHandler {
+
+    private final RuntimeCapability<?> runtimeCapability;
+    private final ServiceController.Mode initialMode;
+    private final ServiceController.Mode embeddedInitialMode;
+
+    HttpAuthenticationFactoryAddHandler(Class<T> serviceType, AttributeDefinition[] attributes, RuntimeCapability<?> runtimeCapability) {
+        this(serviceType, ServiceController.Mode.ACTIVE, attributes, runtimeCapability);
+    }
+
+    HttpAuthenticationFactoryAddHandler(Class<T> serviceType, ServiceController.Mode initialMode, AttributeDefinition[] attributes, RuntimeCapability<?> runtimeCapability) {
+        this(serviceType, initialMode, initialMode, attributes, runtimeCapability);
+    }
+
+    HttpAuthenticationFactoryAddHandler(Class<T> serviceType, ServiceController.Mode initialMode, ServiceController.Mode embeddedInitialMode, AttributeDefinition[] attributes, RuntimeCapability<?> runtimeCapability) {
+        super(new HashSet<>(Collections.singletonList(checkNotNullParam("runtimeCapabilities", runtimeCapability))), attributes);
+        this.runtimeCapability = runtimeCapability;
+        checkNotNullParam("serviceType", serviceType);
+        this.initialMode = checkNotNullParam("initialMode", initialMode);
+        this.embeddedInitialMode = checkNotNullParam("embeddedInitialMode", embeddedInitialMode);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected final void performRuntime(OperationContext context, ModelNode operation, Resource resource)
+            throws OperationFailedException {
+        HttpAuthenticationFactoryService httpAuthenticationFactoryService = new HttpAuthenticationFactoryService();
+
+        ServiceBuilder<T> serviceBuilder = (ServiceBuilder<T>)context.getCapabilityServiceTarget().addCapability(runtimeCapability);
+        serviceBuilder.setInstance(httpAuthenticationFactoryService);
+
+        httpAuthenticationFactoryService.setHttpAuthenticationFactorySupplier(getHttpAuthenticationFactorySupplier(serviceBuilder, context, resource.getModel()));
+
+        installedForResource(commonDependencies(serviceBuilder, dependOnProperties(), dependOnProviderRegistration())
+                .setInitialMode(context.getProcessType() == ProcessType.EMBEDDED_SERVER && context.getRunningMode() == RunningMode.ADMIN_ONLY ? embeddedInitialMode : initialMode)
+                .install(), resource);
+    }
+
+    protected boolean dependOnProperties() {
+        return true;
+    }
+
+    protected boolean dependOnProviderRegistration() {
+        return true;
+    }
+
+    protected void installedForResource(ServiceController<T> serviceController, Resource resource) {}
+
+    protected abstract Callable<HttpAuthenticationFactory> getHttpAuthenticationFactorySupplier(ServiceBuilder<T> serviceBuilder, OperationContext context, ModelNode model) throws OperationFailedException;
```

Comment:
This method (and the class) should not be abstract and the implementation logic starting at https://github.com/wildfly/wildfly-core/pull/3871/files#diff-f20215bf36b61ecbb06008151d2f257cR379 should be moved into this class.

Reply:
An alternative to moving all that logic is to get rid of this class, and change

https://github.com/wildfly/wildfly-core/pull/3871/files#diff-f20215bf36b61ecbb06008151d2f257cR375

to

```
AbstractAddStepHandler add = new BaseAddHandler(HTTP_AUTHENTICATION_FACTORY_RUNTIME_CAPABILITY, attributes) {

            @Override
    protected final void performRuntime(OperationContext context, ModelNode operation, Resource resource)
            throws OperationFailedException {
                   // the performRuntime logic you have here in HttpAuthenticationFactoryAddHandler
     }

     private Callable<HttpAuthenticationFactory> getHttpAuthenticationFactorySupplier(
```

That might be cleanest as it leaves all that code that's in AuthenticationFactoryDefinitions where it is.